File: SDUI/src/sdui/validate.py
"""Local 0.2 profile validation; never load or execute referenced SDL sources."""
from .ast import Document, Reference
from .formatting import fail, formatting

WIDGETS = {
    'button': ('label', {'label': 'string', 'callback': 'reference'}, {'label'}),
    'input': ('text', {'text': 'string', 'value': 'string', 'callback': 'reference'}, {'text'}),
    'svg': ('source', {'source': 'reference', 'label': 'string'}, {'source'}),
}


def widget_arguments(node):
    positional = WIDGETS[node.widget][0]
    return {arg.name or positional: arg.value for arg in node.arguments}


def validate(document: Document) -> None:
    modules, definitions = {}, {}
    for ref in document.references:
        if ref.alias in modules:
            fail('duplicate-module', f'Duplicate module alias {ref.alias}', ref)
        if not ref.path or any(ord(c) < 32 or ord(c) == 127 for c in ref.path):
            fail('module-path', 'Module path must be nonempty without control characters', ref)
        modules[ref.alias] = ref
    for definition in document.definitions:
        if definition.name in definitions or definition.name in modules:
            fail('duplicate-definition', f'Ambiguous definition name {definition.name}', definition)
        definitions[definition.name] = definition
    dependencies = {name: set() for name in definitions}
    for definition in document.definitions:
        names = set()

        def visit(node):
            # Unaliased references take their definition name as the instance name.
            name = node.name or (node.target if node.kind == 'use' else None)
            if name:
                if name in names:
                    fail('duplicate-node', f'Duplicate name {name} in {definition.name}', node)
                names.add(name)
            if node.kind == 'use':
                if node.target not in definitions:
                    fail('unknown-definition', f'Unknown component {node.target}', node)
                dependencies[definition.name].add(node.target)
                # Effective kind/layout checked on the resolved instance below.
            else:
                formatting(node)
            if node.variant and (node.variant not in {'b', 'box'} or node.kind != 'frame'):
                fail('variant', 'Only frame variants box and b are supported', node)
            if node.kind == 'widget':
                validate_widget(node, modules)
            roles, body = set(), False
            for row in node.rows:
                for child in row.items:
                    if node.kind == 'group' and child.kind == 'frame':
                        fail('group-content', 'Frames belong in frame bodies, not widget groups', child)
                    if child.role:
                        if child.role in roles:
                            fail('duplicate-region', f'Duplicate {child.role}', child)
                        roles.add(child.role)
                    else:
                        body = True
                    visit(child)
            if 'body' in roles and body:
                fail('body-conflict', 'Explicit body cannot be mixed with implicit body', node)
        visit(definition.root)
    done = set()

    def cycle(name, stack):
        if name in stack:
            fail('reference-cycle', f'Recursive component reference {name}', definitions[name])
        if len(stack) >= 64:
            fail('depth-limit', 'Reference chain exceeds 64', definitions[name])
        if name not in done:
            for target in sorted(dependencies[name]):
                cycle(target, stack + (name,))
            done.add(name)
    for name in definitions:
        cycle(name, ())
    # Expansion has a separate global budget; references cannot cause exponential work.
    from .normalize import resolve_all
    roots = resolve_all(document)
    widgets = {}

    def collect(instance):
        if instance.kind == 'widget' and not instance.path.rsplit('/', 1)[-1].startswith('$'):
            public_path = '/'.join(part for part in instance.path.split('/') if not part.startswith('$'))
            if public_path in widgets:
                fail('duplicate-instance', f'Ambiguous widget path {public_path}', instance)
            widgets[public_path] = instance
        for row in instance.rows:
            for child in row:
                collect(child)
        for _, region in instance.regions:
            collect(region)
    for root in roots.values():
        collect(root)
    objects, targets = set(), set()
    for connection in document.connections:
        if connection.module not in modules:
            fail('unknown-module', f'Unknown module {connection.module}', connection)
        if connection.definition not in roots:
            fail('unknown-definition', f'Unknown definition {connection.definition}', connection)
        target = '/'.join((connection.definition,) + connection.path)
        if target not in widgets:
            fail('unknown-widget', f'Unknown widget instance {target}', connection)
        obj = (connection.module, connection.object)
        if obj in objects or target in targets:
            fail('duplicate-connection', 'One connection per SDL object and widget', connection)
        objects.add(obj)
        targets.add(target)


def validate_widget(node, modules):
    if node.widget not in WIDGETS:
        fail('widget-kind', f'Unsupported widget kind {node.widget}; quoted text is Markdown', node)
    positional, fields, required = WIDGETS[node.widget]
    args, named_seen = {}, False
    for arg in node.arguments:
        if arg.name is None and (args or named_seen):
            fail('argument-order', 'Only one positional argument before named arguments', arg)
        key = arg.name or positional
        named_seen |= arg.name is not None
        if key in args:
            fail('duplicate-argument', f'Duplicate widget property {key}', arg)
        if key not in fields:
            fail('widget-property', f'Unknown {node.widget} property {key}', arg)
        value, expected = arg.value, fields[key]
        if isinstance(value, Reference):
            if expected != 'reference':
                fail('argument-type', f'{key} requires {expected}', arg)
            if value.module not in modules:
                fail('unknown-module', f'Unknown module {value.module}', value)
        elif expected == 'reference' or value.kind != expected:
            fail('argument-type', f'{key} requires {expected}', arg)
        args[key] = value
    if node.name is None and 'callback' in args:
        fail('widget-name', 'A callback requires a named widget', node)
    if required - args.keys():
        fail('missing-argument', f'Missing properties {sorted(required - args.keys())}', node)

File: SDUI/go/parser/validate.go
package parser

import "strings"

type widgetSchema struct {
	position string
	fields   map[string]string
	required string
}

var widgets = map[string]widgetSchema{
	"button": {"label", map[string]string{"label": "string", "callback": "reference"}, "label"},
	"input":  {"text", map[string]string{"text": "string", "value": "string", "callback": "reference"}, "text"},
	"svg":    {"source", map[string]string{"source": "reference", "label": "string"}, "source"},
}

func widgetArguments(n *Node) map[string]any {
	out := map[string]any{}
	for _, a := range n.Arguments {
		k := val(a.Name)
		if a.Name == nil {
			k = widgets[val(n.Widget)].position
		}
		out[k] = a.Value
	}
	return out
}
func validateWidget(n *Node, modules map[string]bool) {
	schema, ok := widgets[val(n.Widget)]
	if !ok {
		fail("widget-kind", "Unsupported widget "+val(n.Widget), n.Span)
	}
	seen := map[string]bool{}
	for _, a := range n.Arguments {
		if a.Name == nil && len(seen) > 0 {
			fail("argument-order", "Only one positional argument before named arguments", a.Span)
		}
		k := val(a.Name)
		if a.Name == nil {
			k = schema.position
		}
		if seen[k] {
			fail("duplicate-argument", "Duplicate "+k, a.Span)
		}
		expected, ok := schema.fields[k]
		if !ok {
			fail("widget-property", "Unknown widget property "+k, a.Span)
		}
		switch v := a.Value.(type) {
		case Reference:
			if expected != "reference" {
				fail("argument-type", "Expected "+expected, a.Span)
			}
			if !modules[v.Module] {
				fail("unknown-module", "Unknown module "+v.Module, v.Span)
			}
		case Literal:
			if v.Kind != expected {
				fail("argument-type", "Expected "+expected, a.Span)
			}
		default:
			fail("argument-type", "Invalid argument", a.Span)
		}
		seen[k] = true
	}
	if seen["callback"] && n.Name == nil {
		fail("widget-name", "A callback requires a named widget", n.Span)
	}
	if !seen[schema.required] {
		fail("missing-argument", "Missing "+schema.required, n.Span)
	}
}
func validateLocal(d *Document) {
	if d == nil {
		fail("document", "Nil document", Span{})
	}
	modules := map[string]bool{}
	definitions := map[string]Definition{}
	for _, r := range d.References {
		if modules[r.Alias] {
			fail("duplicate-module", "Duplicate module", r.Span)
		}
		bad := r.Path == ""
		for _, c := range r.Path {
			bad = bad || c < 32 || c == 127
		}
		if bad {
			fail("module-path", "Invalid module path", r.Span)
		}
		modules[r.Alias] = true
	}
	for _, def := range d.Definitions {
		if _, ok := definitions[def.Name]; ok || modules[def.Name] {
			fail("duplicate-definition", "Ambiguous definition "+def.Name, def.Span)
		}
		definitions[def.Name] = def
	}
	deps := map[string][]string{}
	for _, def := range d.Definitions {
		names := map[string]bool{}
		var visit func(*Node)
		visit = func(n *Node) {
			name := val(n.Name)
			if name == "" && n.Kind == "use" {
				name = val(n.Target)
			}
			if name != "" {
				if names[name] {
					fail("duplicate-node", "Duplicate name "+name, n.Span)
				}
				names[name] = true
			}
			if n.Kind == "use" {
				if _, ok := definitions[val(n.Target)]; !ok {
					fail("unknown-definition", "Unknown component "+val(n.Target), n.Span)
				}
				deps[def.Name] = append(deps[def.Name], val(n.Target))
			} else {
				formatting(n, n.Kind, len(n.Rows))
			}
			if n.Variant != nil && (n.Kind != "frame" || !member(*n.Variant, "b box")) {
				fail("variant", "Only frame variants box/b supported", n.Span)
			}
			if n.Kind == "widget" {
				validateWidget(n, modules)
			}
			roles := map[string]bool{}
			body := false
			for _, row := range n.Rows {
				for _, child := range row.Items {
					if n.Kind == "group" && child.Kind == "frame" {
						fail("group-content", "Frame in widget group", child.Span)
					}
					if child.Role != nil {
						if roles[*child.Role] {
							fail("duplicate-region", "Duplicate region", child.Span)
						}
						roles[*child.Role] = true
					} else {
						body = true
					}
					visit(child)
				}
			}
			if roles["body"] && body {
				fail("body-conflict", "Explicit and implicit body cannot mix", n.Span)
			}
		}
		visit(def.Root)
	}
	done := map[string]bool{}
	stack := map[string]bool{}
	var cycle func(string, int)
	cycle = func(name string, depth int) {
		if stack[name] {
			fail("reference-cycle", "Recursive reference "+name, definitions[name].Span)
		}
		if depth >= 64 {
			fail("depth-limit", "Reference chain exceeds 64", definitions[name].Span)
		}
		if done[name] {
			return
		}
		stack[name] = true
		for _, child := range deps[name] {
			cycle(child, depth+1)
		}
		delete(stack, name)
		done[name] = true
	}
	for _, def := range d.Definitions {
		cycle(def.Name, 0)
	}
}
func validateConnections(d *Document, roots map[string]*Instance) {
	modules := map[string]bool{}
	for _, r := range d.References {
		modules[r.Alias] = true
	}
	paths := map[string]bool{}
	var visit func(*Instance)
	visit = func(n *Instance) {
		parts := strings.Split(n.Path, "/")
		if n.Kind == "widget" && !strings.HasPrefix(parts[len(parts)-1], "$") {
			public := []string{}
			for _, p := range parts {
				if !strings.HasPrefix(p, "$") {
					public = append(public, p)
				}
			}
			path := strings.Join(public, "/")
			if paths[path] {
				fail("duplicate-instance", "Ambiguous path "+path, n.Span)
			}
			paths[path] = true
		}
		for _, row := range n.Rows {
			for _, child := range row {
				visit(child)
			}
		}
		for _, reg := range n.Regions {
			visit(reg.Node)
		}
	}
	for _, def := range d.Definitions {
		visit(roots[def.Name])
	}
	objects, targets := map[string]bool{}, map[string]bool{}
	for _, con := range d.Connections {
		if !modules[con.Module] {
			fail("unknown-module", "Unknown module "+con.Module, con.Span)
		}
		if roots[con.Definition] == nil {
			fail("unknown-definition", "Unknown definition", con.Span)
		}
		target := con.Definition + "/" + strings.Join(con.Path, "/")
		if !paths[target] {
			fail("unknown-widget", "Unknown widget "+target, con.Span)
		}
		obj := con.Module + "." + con.Object
		if objects[obj] || targets[target] {
			fail("duplicate-connection", "One connection per object/widget", con.Span)
		}
		objects[obj] = true
		targets[target] = true
	}
}

// Validate checks local semantics including bounded instance expansion.
func Validate(d *Document) error { _, err := Normalize(d); return err }
